### ISSUE TITLE
Overly strict behavior of definite assignment

### ISSUE THREAD

**Issue by davidcok** [dafny-lang/dafny#3232]
### Dafny version

3.10.0+dev

### Code to produce this issue

```dafny
method Enqueue(n: nat)
  {
    var a := new int[n];
    forall i | 0 <= i < n {
      a[i] := i;
    }
  }
```

### Command to run and resulting output

```code
daffy verify test.dfy
```

### What happened?

/Users/davidcok/projects/dafny/dafny/docs/DafnyRef/test.dfy(3,13): Error: unless an initializer is provided for the array elements, a new array of 'int' must have empty size

It appears that the default setting of definite assignment requires array elements to be set on allocation even if they aren't used. This is a real nuisance and requires either globally using --relax-definite-assignment or putting in an unneeded initializer, which is unnecessary computation.

### What type of operating system are you experiencing the problem on?

Mac

**Comment by RustanLeino** [comment]
I believe this was an unintended nuisance. The reason for tightening up the rules about definite assignment was to try to reduce easy bugs and confusion. For example, a method that accidentally omits an assignment to an out-parameter would get complaints about the postcondition not holding, whereas a simple message saying the out-parameter was never assigned is likely to more to the point for the user.

For an array allocation, continuing to use the (nondeterministic) auto-init of elements seems desirable.

I suspect the previous `/definiteAssignment` switch probably controlled both of these aspects. I think initialization of arrays should work like it did before under `/definiteAssignment:1` (unless the user asks of Dafny to restrict the use of nondeterministic features).

**Comment by keyboardDrummer** [comment]
I think ideally the error is only given if the non-deterministic array elements are read before they are assigned. Can we not implement that @RustanLeino ?

Alternatively, we could also allow array initialisation with a constant, so you could do 

```dafny
method Enqueue(n: nat)
  {
    var a := new int[n](0);
    forall i | 0 <= i < n {
      a[i] := i;
    }
  }
```

which in the compilation is translated to some fast array initialisation, to enable easily working around the issue with little/no performance penalty.

**Comment by RustanLeino** [comment]
Just for the record, to check that array elements have been initialized before they are used requires detailed theorem proving. The definite-assignment checks we do for other variables uses the theorem prover in a simpler way.

**Comment by davidcok** [comment]
Yes. but in simple cases -- where all elements are pretty clearly initialized -- which is the common case -- it ought to be any easy proof. If only some elements are initialized, the user might forgive the prover for not knowing.
